Mathlib/Order/BourbakiWitt.lean

```diff
@@ -0,0 +1,227 @@
+/-
+Copyright (c) 2025 Finn Mortimore. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Finn Mortimore
+-/
+import Mathlib.Order.Preorder.Chain
+import Mathlib.Data.Set.Lattice
+import Mathlib.Dynamics.FixedPoints.Basic
+
+/-!
+# Bourbaki-Witt Theorem
+
+This file proves the Bourbaki-Witt Theorem.
+
+## Main definitions
+
+- class `ChainCompletePartialOrder` : A non-empty partial order is a chain complete partial order
+  such that every non-empty chain has a supremum
+- structure `Inflationary` : Let $X$ be a partial order. A function $f : X → X$ is inflationary if,
+  for all $x ∈ X$, $x ≤ f x$
+
+## Main statements
+
+- `bourbaki_witt` : The Bourbaki-Witt Theorem : If $X$ is a chain complete
+partial order and $f : X → X$ is inflationary, then $f$ has a fixed point
+
+## References
+
+The proof used can be found in [serge_lang_algebra]
+-/
+
+/-- The type of non-empty chains of an order -/
+@[ext]
+structure NonemptyChain (α : Type) [LE α] where
+  /-- The underlying set of a non-empty chain -/
+  carrier : Set α
+  /-- By definition, a non-empty chain is non-empty -/
+  Nonempty' : carrier.Nonempty
+  /-- By definition, a non-empty chain is a chain -/
+  isChain' : IsChain (· ≤ ·) carrier
+
+instance {α : Type} [LE α] : SetLike (NonemptyChain α) α where
+  coe := NonemptyChain.carrier
+  coe_injective' _ _ := NonemptyChain.ext
+
+/-- A chain complete partial order is a non-empty partial order such that every chain has a supremum
+(which we call `cSup`) -/
+class ChainCompletePartialOrder (α : Type) [Nonempty α] extends PartialOrder α where
+  /-- The supremum of a non-empty chain -/
+  cSup : NonemptyChain α → α
+  /-- `cSup` is an upper bound of the non-empty chain -/
+  le_cSup (c : NonemptyChain α) (x : α) : x ∈ c.carrier → x ≤ cSup c
+  /-- `cSup` is a lower bound of the set of upper bounds of the non-empty chain -/
+  cSup_le (c : NonemptyChain α) (x : α) : (∀ y ∈ c.carrier, y ≤ x) → cSup c ≤ x
+
+/-- A function `f : α → α` is inflationary if for all `x : α`, `x ≤ f x` -/
+structure Inflationary (α : Type) [PartialOrder α] where
+  /-- The underlying function of an inflationary function -/
+  toFun : α → α
+  /-- The underlying function of an inflationary function is inflationary -/
+  le_map (x : α) : x ≤ toFun x
+
+instance {α : Type} [PartialOrder α] : CoeFun (Inflationary α) (fun _ ↦ α → α) where
+  coe := Inflationary.toFun
+
+open ChainCompletePartialOrder Set
+
+section
+
+namespace BourbakiWitt
+
+variable {α : Type} [Nonempty α] [ChainCompletePartialOrder α] {x : α} {f : Inflationary α}
+
+/-- The definition of an admissible set with base point `x` and inflationary function `f` -/
```

Comment:
A docstring should ideally give some explanation of what a definition means, without needing to go to the definition to read the Lean code. So something like:
```suggestion
/-- An admissible set for given `x` and `f` has `x` as a least element and is closed under applying `f` and `cSup`. -/
```